### PR TITLE
WIP Set toleration on dns-controller

### PR DESCRIPTION
**Pull request by justinsb**
But only for >= k8s 1.6.  Unclear if our version of kubectl can actually
do this!

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2033)

### CODE REVIEW

**Comment by justinsb** [comment]
This isn't actually needed until we start setting the taints

**Comment by kris-nova** [comment]
@justinsb rebase please :)

File: nodeup/pkg/model/kubelet.go [hunk 1]
```diff
@@ -242,22 +246,10 @@ func (b *KubeletBuilder) buildKubeconfig() (string, error) {
...
-		glog.Warningf("Building a synthetic instance group")
-		instanceGroup = &kops.InstanceGroup{}
-		instanceGroup.ObjectMeta.Name = "synthetic"
-		if b.IsMaster {
-			instanceGroup.Spec.Role = kops.InstanceGroupRoleMaster
-		} else {
-			instanceGroup.Spec.Role = kops.InstanceGroupRoleNode
-		}
-		//b.InstanceGroup = instanceGroup
+	if b.InstanceGroup == nil {
```
Comment (kris-nova) [inline nodeup/pkg/model/kubelet.go]: Does this break backwards compatibility?

File: nodeup/pkg/model/protokube.go [hunk 2]
```diff
@@ -178,5 +180,9 @@ func (t *ProtokubeBuilder) ProtokubeFlags(k8sVersion semver.Version) *ProtokubeF
 
 	f.DNSInternalSuffix = fi.String(".internal." + t.Cluster.ObjectMeta.Name)
 
+	if k8sVersion.Major == 1 && k8sVersion.Minor <= 5 {
```
Comment (kris-nova) [inline nodeup/pkg/model/protokube.go]: less than or equal to 5 gets taints? I thought taints was a new thing to 1.6

File: pkg/apis/kops/validation/instancegroup.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+func ValidateInstanceGroup(g *kops.InstanceGroup) error {
+	if g.ObjectMeta.Name == "" {
+		return field.Required(field.NewPath("Name"), "")
+	}
+
+	if g.Spec.Role == "" {
+		return field.Required(field.NewPath("Role"), "Role must be set")
+	}
+
+	switch g.Spec.Role {
```
Comment (kris-nova) [inline pkg/apis/kops/validation/instancegroup.go]: This switch doesn't seem to be doing anything (it should always hit default).

Just in place for future logic? Or..?

File: pkg/apis/kops/validation/instancegroup.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	if k8sVersion.Major == 1 && k8sVersion.Minor <= 5 {
+		if len(g.Spec.Taints) > 0 {
+			if !(g.IsMaster() && g.Spec.Taints[0] == kops.TaintNoScheduleMaster15 && len(g.Spec.Taints) == 1) {
+				allErrs = append(allErrs, field.Invalid(fieldPath.Child("Spec").Child("Taints"), g.Spec.Taints, "User-specified taints are not supported before kubernetes version 1.6.0"))
+			}
+		}
+	}
+
+	if len(allErrs) != 0 {
+		return allErrs[0]
```
Comment (kris-nova) [inline pkg/apis/kops/validation/instancegroup.go]: Question:

Why track all the errors, if we are only returning index 0? Seems like we would want to chain all the errors together here to know what's up?

```go
if len(allErrs) != 0 {
    err := fmt.Printf("errors while cross validating InstanceGroup: ")
    for i, err := range allErrs {
        err = fmt.Printf("%i (%v)", i, v)
    }
    return err
}
```

**Comment by justinsb** [comment]
Merged separately :-)

**Pull request closed without merging**